[processor/tailsampling] Include componentID as prefix in metrics 'policy'

EOjeah opened the pull request:
**Description:**
Fixing a bug - This change includes the componentID as a dot prefix to the metrics `policy` dimension when generating metrics for the processor. The change ensures that similarly named policy's in the tail sampling processor that belong to different components also has a unique value in the `policy` field for the metrics.

Also includes minor refactor change to rename `telemetry` to `telemetryBuilder` where applicable (return type == `NewTelemetryBuilder`)

Resolves: #34099

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** 
Ran the collector locally with `make run` with the configuration below which uses the tail sampling processor and has metrics exposed in prometheus format. Sending sample zipkin spans to the receiver 
```yaml
receivers:
  zipkin:

processors:
  tail_sampling:
    policies:
      [
          {
            name: test-policy-1,
            type: always_sample
          }
      ]

  tail_sampling/custom_name:
    policies:
      [
          {
            name: test-policy-1,
            type: always_sample
          }
      ]

exporters:
  debug:

service:
  telemetry:
    logs:
    metrics:
  pipelines:
    traces:
      receivers: [zipkin]
      processors: [tail_sampling, tail_sampling/custom_name]
      exporters: [debug]
```

Curling the metrics endpoint shows the policy name is unique for both tail sampling processors
```bash
otelcol_processor_tail_sampling_sampling_decision_latency_bucket{policy="custom_name.test-policy-1",service_instance_id="X",service_name="otelcontribcol",service_version="0.105.0-dev",le="5000"} 1

otelcol_processor_tail_sampling_sampling_decision_latency_bucket{policy="test-policy-1",service_instance_id="X",service_name="otelcontribcol",service_version="0.105.0-dev",le="5000"} 1
```

Tasks

- [x] Confirm prefix separator as `.` 
- [x] Update change log entry

EOjeah committed:
Include componentID as prefix in metrics 'policy'

This change includes the componentID as a dot prefix to the metrics
'policy' dimension when generating metrics for the processor. The change
ensures that similarly named policys in the tail sampling processor that
belong to different components also has a unique value.

Resolves: #34099

EOjeah commented:
Not urgent but tagging @jpkrohling (currently on holiday) just so it's not closed due to inactivity

jpkrohling reviewed (commented):
There seems to be a lot of unrelated changes in this PR, like the renaming of telemetry to telemetryBuilder. The actual change is quite small and localized, and reviewing it would have been otherwise straightforward. 

I have not seen new tests exercising the new code path: can you please add a new test?

EOjeah committed:
Include Test to varify policy dimension value

Undo previous rename of telemtry to telemetryBuilder as it's not related to the attached issue. In order for the component ID to be passed from settings, the newTraceProcessor definition had to be chanaged to accept processor.Settings and not just the TelemetrySettings

EOjeah committed:
Merge branch 'main' of github.com:EOjeah/opentelemetry-collector-contrib into bug/unique-policy-name-tail-sampling-processor

EOjeah commented:
Reverted the change that renamed telemetry to telemetryBuilder. 

In order to receive the processor settings, which contains the `component.ID` in the `newTracesProcessor` function, the newTracesProcessor function had to be updated to accept the `processor.Settings` and not just the `processor.Settings.TelemetrySettings` 

### [Settings](https://pkg.go.dev/go.opentelemetry.io/collector/processor@v0.107.0/internal#Settings)
```go
// Settings is passed to Create* functions in Factory.
type Settings struct {
  // ID returns the ID of the component that will be created.
  ID component.ID

  component.TelemetrySettings

  // BuildInfo can be used by components for informational purposes
  BuildInfo component.BuildInfo
}
```

Included a test that verifies the policy value in the metrics generated includes the ID as a dot `.` prefix

jpkrohling reviewed (commented):
The actual change LGTM, just not sure about the reason for renaming the telemetry settings.

jpkrohling commented on processor/tailsamplingprocessor/processor.go:
```diff
@@ -83,8 +83,9 @@ type Option func(*tailSamplingSpanProcessor)
-func newTracesProcessor(ctx context.Context, settings component.TelemetrySettings, nextConsumer consumer.Traces, cfg Config, opts ...Option) (processor.Traces, error) {
-	telemetry, err := metadata.NewTelemetryBuilder(settings)
+func newTracesProcessor(ctx context.Context, set processor.Settings, nextConsumer consumer.Traces, cfg Config, opts ...Option) (processor.Traces, error) {
+	telemetrySettings := set.TelemetrySettings
```
No, it's okay. This is small enough to not be a problem.

EOjeah committed:
Merge branch 'main' of github.com:EOjeah/opentelemetry-collector-contrib into bug/unique-policy-name-tail-sampling-processor

EOjeah committed:
Include change log entry

EOjeah committed:
Merge branch 'main' of github.com:EOjeah/opentelemetry-collector-contrib into bug/unique-policy-name-tail-sampling-processor

EOjeah committed:
Merge branch 'main' of github.com:EOjeah/opentelemetry-collector-contrib into bug/unique-policy-name-tail-sampling-processor

EOjeah commented:
added the changelog entry so re-request approval

EOjeah committed:
Merge branch 'main' of github.com:EOjeah/opentelemetry-collector-contrib into bug/unique-policy-name-tail-sampling-processor
